### ISSUE TITLE
log_ratio problem 

### ISSUE THREAD

**Issue by ecada** [tristandeleu/pytorch-maml-rl#24]
Hello,
Thank you for sharing this wonderful implementation. In pytorch-maml-rl/maml_rl/metalearner.py line 142, when we are calculating the log ratio (which is essentially importance sampling) why do we put the old_pi in the denominator of the ratio? The advantages are sampled from pi, not old_pi so isn't the importance weight implementation wrong?

**Comment by tristandeleu** [comment]
Hi, thank you for the kind words!
This ratio comes from how the surrogate loss is defined in TRPO (for example, see this [implementation of TRPO](https://github.com/openai/baselines/blob/master/baselines/trpo_mpi/trpo_mpi.py#L196)). However the advantages are indeed computed with samples from `old_pi`. `old_pi` is actually the policy we are updating, and `pi` is a "candidate" policy we could chose for the update (this may be an unfortunate naming convention). But you are right, there is kind of a distribution mismatch happening here, but this is an approximation that is done in TRPO. [This great set of slides](http://rail.eecs.berkeley.edu/deeprlcourse/static/slides/lec-9.pdf) explains all this in details.

**Comment by ecada** [comment]
Thank you for your kind reply! 
I've gone through the TRPO implementation and the slides and what they do exactly is to assign the old_pi with the new_pi before taking any gradient steps. Hence the advantages are sampled from the old_pi so it makes sense because the old_pi is assigned with the new_pi that the advantages are sampled from. That is the reason the first ratio is 1 before taking gradient descent. In the 13th slide of the set it's also seen that the advantages are sampled from the old_pi and the gradient is computed with respect to the new_pi which we haven't yet sampled from.
